Exploring GraphiQL: The In-Browser IDE for GraphQL (#338)
tailcallhq/tailcallhq.github.io: outlines/exploring-graphiql-the-in-browser-ide-for-graphql.md

# Exploring GraphiQL: The In-Browser IDE for GraphQL

## Introduction

A. Definition and overview of GraphiQL

## Why use GraphiQL?

A. Advantages/benefits of using GraphiQL

## Accessing GraphiQL

For this guide, accessing GraphiQL can be done using any of the methods below:

A. Simply head over to [https://tailcall.run/playground/](https://tailcall.run/playground/) and give your graphql server url in input. Note, make sure your graphql server allows cors for Tailcall domain

B. After starting your [GraphQL server](https://tailcall.run/docs/#starting-the-graphql-server), head over to the GraphiQL playground link you get

## Setting up GraphiQL

A. Configuring GraphiQL

B. Connecting GraphiQL to your Tailcall GraphQL server

## Exploring the GraphiQL Interface

A. Exploring the GraphiQL interface and understanding the functions of the various panes.

B. Explain each of the main parts in the GraphiQL interface

- Query Editor
- Variables Editor
- Response pane
- Documentation Explorer.

> 

Reviewer: How to add headers?

Author: Do you mean the parts I added "head to"?

Reviewer: How to add headers? in GraphiQL

Author: Alright, I'd add that.

(screenshot each of this interface)

E. Test with variables: Add variables in the Variables Editor to make your query dynamic.

```graphql
# query showing the variable added
```

(screenshot the variable pane)

F. Analysing the response

- Inspect the response in the Result Pane to ensure you're getting the expected data

```json
// sample of successful response
```

- Explain an instance where there is an error and how users can use the error messages to identify the issues.

```json
// sample of an error response
```

### Debugging Strategies

A. Errors that can lead to bugs in the Queries

- Syntax

(Add a screenshot of a query in the editor with a syntax error)

- Type Mismatches

(Add a screenshot of a query in the editor with a type mismatch error)

- Field Resolution issues.

(Add a diagram analysing the field resolution issue)

B. Error Messages

- Understanding and interpreting GraphQL error messages

C. Network Tab in Developer Tools

- Analyzing network requests and responses for additional insights.

> ### Best Practices

Reviewer: Not relavant

Author: Alright, I'd remove the whole thing.

Author: > Not relavant

Here, I removed all the sections from "Key Features and Capabilities of GraphiQl" to "Debugging Strategies"

A. Writing clear and well-structured queries.

B. Using modular queries, i.e. breaking down complex queries into smaller and manageable chunks.

C. Implement thorough/consistent query testing strategies to detect errors early.

D. Adding in-line comments in your Query

## Alternatives to GraphiQL

> A. Explain that there are also alternatives to GraphiQL. There are other GraphQL playgrounds like Insomnia, Altair client and so on

Reviewer: add screenshots for each alternative

B. Add a table to compare GraphiQl to Altair Client and Insomnia

## Conclusion

A. Summary of what we discussed

B. Final thoughts on the subject

> ## Resources

Reviewer: Drop this section

Author: Alright, I will.

Author: > Drop this section

Including the conclusion?

Reviewer: ## Resources section

Author: > ## Resources section

Done

Reviewer: its still there

Author: > its still there

I have resolved it in this new commit. Thank you!

# Exploring GraphiQL: The In-Browser IDE for GraphQL

**Target Audience:** This article is for all developers and testers who integrate or test GraphQL APIs, whether front-end developers, back-end developers, or QA testers. It's also for anyone else interested in the world of GraphQL APIs.

**Aim:** This article aims to provide developers with an in-depth understanding of GraphiQL, its features, and how to effectively use it to interact with GraphQL APIs. It will cover the fundamentals, and practical applications for utilizing GraphiQL as an in-browser IDE for GraphQL.

## Introduction

> Why waste time and effort troubleshooting your GraphQL API directly in your application when there's a more efficient way? Imagine testing and perfecting your queries outside your app, ensuring they work flawlessly before implementation. Sounds ideal, right? With [GraphiQL](https://az.dev/swapi-graphql), you're saving time and gaining control over your development process, empowering you to work more confidently and effectively.

Reviewer: ```suggestion
Why waste time and effort troubleshooting your GraphQL API directly in your application when there's a more efficient way? Imagine testing and perfecting your queries outside your app, ensuring they work flawlessly before implementation. Sounds ideal, right? With [GraphiQL](https://tailcall.run/playground/), you're saving time and gaining control over your development process, empowering you to work more confidently and effectively.
```

## Prerequisites
To follow along with this guide, you need:
* Fundamental understanding of GraphQL 
* Basic knowledge of schemas, queries, mutations, and resolvers. 
* A web browser
* A GraphQL API endpoint: This is how GraphiQL can communicate with a server to access the data. 

## What is GraphiQL?
* Definition and overview of GraphiQL

## Why use GraphiQL?
* Advantages/benefits of using GraphiQL (things like Improving your request, validating your improvements, and easily debugging requests running into problems)

## Methods for accessing GraphiQL
To access GraphiQL, you are not limited to the default in-browser method. 
* Head to [https://graphql.org/swapi-graphql](https://graphql.org/swapi-graphql) by querying your search engine (browser) to access the GraphiQL online editor, which works with the Star Wars data, and it is publicly available for you to use

> * If you are using Gatsby, you can start the Gatsby development server by pasting this: `http://localhost:8000/___graphql` into your browser

Reviewer: ```suggestion
* If you are using Gatsby, you can start the Gatsby development server by pasting this: `http://localhost:8000/___graphql` into your browser
```
This should reference tailcall

Author: @amitksingh1490 How do I do this? Is it the Tailcall website I am referencing?

Reviewer: check this out https://tailcall.run/docs/#starting-the-graphql-server

* Using Shopify via the [Shopify GraphiQl app](https://shopify-graphiql-app.shopifycloud.com/login)

> * Using your own API endpoint: just type your GraphQL server endpoint URL into your browser

Reviewer: use https://tailcall.run/playground/

## Getting Started with the GraphiQL
* Configuring GraphiQL

> * Connecting GraphiQL to your GraphQL server

Reviewer: ```suggestion
* Connecting GraphiQL to your Tailcall GraphQL server
```

## Exploring the GraphiQL Interface
* Exploring the GraphiQL interface and understanding the functions of the various panes. The application is two-pane by default, but it becomes three when we access the Documentation Explorer.
* Explain each of the main parts in the GraphiQL interface: Query Editor, Variables Editor, Response pane, and Documentation Explorer.

## Key Features and Capabilities of GraphiQl
Explain the various core features that GraphiQL provides.
* Type-ahead and Auto Completion feature that is aware of the GraphQL type schema you are currently exploring. In addition, explain Context-awareness.
* Live syntax and validation error highlighting.
* Query history
* Fragments: This feature helps us maintain the Don’t Repeat Yourself (DRY) pattern when dealing with repetitive code in our query. Instead, you can easily merge fragments into a query.

> * Documentation: GraphiQL generates documentation for your schema, making it easy to understand and explore.

Reviewer: screenshot for documentation How it looks

## Testing out the Interface
Writing and executing Queries, mutations and Subscriptions
* Writing the query
* Integrating Variables with Your Query
* Utilising fragments to aid query construction

> * Executing/Testing the query: Sending requests and handling responses

Reviewer: screen shots

## Prerequisites
To follow along with this guide, you need:
* Fundamental understanding of GraphQL 
* Basic knowledge of schemas, queries, mutations, and resolvers. 
* A web browser
* A GraphQL API endpoint: This is how GraphiQL can communicate with a server to access the data. 

## What is GraphiQL?
* Definition and overview of GraphiQL

## Why use GraphiQL?
* Advantages/benefits of using GraphiQL (things like Improving your request, validating your improvements, and easily debugging requests running into problems)

## Methods for accessing GraphiQL
To access GraphiQL, you are not limited to the default in-browser method. 
* Head to [https://tailcall.run/playground/?u=http://127.0.0.1:8000/graphql](https://tailcall.run/playground/?u=http://127.0.0.1:8000/graphql) by querying your search engine (browser) to access the GraphiQL online editor, which works with the Star Wars data, and it is publicly available for you to use

> * If you are using Gatsby, you can start the Gatsby development server by referencing this: (https://tailcall.run/docs/#starting-the-graphql-server)[https://tailcall.run/docs/#starting-the-graphql-server]

Reviewer: > start the Gatsby development server

Doesn't make sense

Author: Alright. But is the first point okay?

Author: > > start the Gatsby development server
> 
> Doesn't make sense

Alright. I'd just remove that and add the part of starting a server from the link you shared earlier.

Reviewer: >which works with the Star Wars data, and it is publicly available for you to use 

is incorrect

Author: I'd just remove that part. The initial link supports the Star Wars data.

Author: > > which works with the Star Wars data, and it is publicly available for you to use
> 
> is incorrect

To clarify is there more than one method to access GraphiQL when using Tailcall?

Reviewer: when you start the Tailcall server you get a link to playground.
or you can directly go to the https://tailcall.run/playground/ and give your graphql server url in input
Only thing you need to make sure your graphql server should allow cors for Tailcall domain

Author: > when you start the Tailcall server you get a link to playground. or you can directly go to the https://tailcall.run/playground/ and give your graphql server url in input Only thing you need to make sure your graphql server should allow cors for Tailcall domain

Alright, now I get it. Thank you for clarifying. I have committed the changes.
